### PR TITLE
Support environment variables in command

### LINKED ISSUE #3406: Support environment variables for a specific command
/kind user-story

## User Story
As a stack creator, I want to specify environment variables in a command in the devfile so that the environment variables are used when executing a specific command

## Acceptance Criteria 
- [ ] It should support the env definition in a command as specified to the devfile 2.0 spec: https://github.com/devfile/kubernetes-api/issues/21

## Links 
- Related issue: https://github.com/devfile/kubernetes-api/issues/21

/kind user-story
/area devfile

### PR DESCRIPTION
**Pull request by yangcao77**
**What type of PR is this?**
/kind user-story
/area devfile

**What does does this PR do / why we need it**:
This PR  allows to specify environment variables in a command in the devfile so that the environment variables are used when executing the command.

**Which issue(s) this PR fixes**:

Fixes #3406 

**How to test changes / Special notes to the reviewer**:

**automation:** 
run `make test-cmd-devfile-push`

**manual:** 
Use the devfile.yaml within this PR: `devfile-with-multiple-defaults.yaml`

run `odo create` and `odo push --build-command firstbuild --run-command singleenv`
`kubectl exec -it <pod> ls /projects/nodejs-web-app/` should see `test_env_variable` directory exist

### CODE REVIEW

File: pkg/devfile/adapters/kubernetes/utils/utils.go [hunk 1]
```diff
@@ -156,10 +156,19 @@ func UpdateContainersWithSupervisord(devfileObj devfileParser.DevfileObj, contai
...
 			if !isEnvPresent(container.Env, adaptersCommon.EnvOdoCommandRun) {
 				klog.V(4).Infof("Updating container %v env with run command", container.Name)
+				var setEnvVariable, command string
+				for _, envVar := range runCommand.Exec.Env {
+					setEnvVariable = setEnvVariable + fmt.Sprintf("%v=\"%v\" ", envVar.Name, envVar.Value)
+				}
+				if setEnvVariable == "" {
+					command = runCommand.Exec.CommandLine
+				} else {
+					command = setEnvVariable + " && " + runCommand.Exec.CommandLine
```
Comment (mik-dass) [inline pkg/devfile/adapters/kubernetes/utils/utils.go:166]: this leaves a extra space between the last envs and the command
```
ODO_COMMAND_RUN=JAVA_OPTS="some value"  && npm start
```

File: tests/examples/source/devfiles/nodejs/devfile-with-multiple-defaults.yaml [hunk 2]
```diff
@@ -51,5 +51,27 @@ commands:
...
+        kind: run
+  - exec:
+      id: multipleenv
+      component: runtime
+      commandLine: "mkdir $ENV1 $ENV2"
+      env:
+       - name: ENV1
+         value: "test_env_variable1"
+       - name: ENV2
+         value: "test_env_variable2"
```
Comment (mik-dass) [inline tests/examples/source/devfiles/nodejs/devfile-with-multiple-defaults.yaml:74]: can we have a env var with a space too?

File: pkg/devfile/adapters/kubernetes/utils/utils.go [hunk 3]
```diff
@@ -156,10 +156,19 @@ func UpdateContainersWithSupervisord(devfileObj devfileParser.DevfileObj, contai
 			// This is done, so supervisord can use it in it's program
 			if !isEnvPresent(container.Env, adaptersCommon.EnvOdoCommandRun) {
 				klog.V(4).Infof("Updating container %v env with run command", container.Name)
+				var setEnvVariable, command string
```
Comment (mik-dass) [inline pkg/devfile/adapters/kubernetes/utils/utils.go:159]: Please update the unit test for this function.

**Comment by dharmit** [comment]
@yangcao77 can you please also share a working example of this that would help test the functionality of this PR? If it's already there at some place (code/issue), please point me to it.

**Comment by dharmit** [comment]
AFAICS, there's not make target by the specified name. I hope I'm not doing something really stupid here. :cry: 

> [](https://github.com/openshift/odo/issues/3406)
> 
> **How to test changes / Special notes to the reviewer**:
> 
> automation:
> run `make test-push-devfile-test`

```sh
$ git checkout upstream/pr/3537
Previous HEAD position was f7c51a7b0 Refactor odo delete logic
HEAD is now at 9b7cc15dc Merge branch 'master' of github.com:openshift/odo into 3406-environmentVariable

$ make test-push-devfile-test
make: *** No rule to make target 'test-push-devfile-test'.  Stop.

$ git log --oneline -n7
9b7cc15dc (HEAD, upstream/pr/3537) Merge branch 'master' of github.com:openshift/odo into 3406-environmentVariable
4d69aab92 added deprecation and error messages for 'odo update' (#3453)
8791263cf Fixing kubeconfig path and login session for 4.x cluster (#3528)
72b44d7bb Adds a error message for creation of routes with host flag (#3541)
c7d7fcb16 Update odo delete for no component and devfile.yaml (#3405)
42fe69318 used MatchJSON in tests wherever needed (#3523)
941bcf83b catalog search services honor experiment flag (#3366)

$ make && cp -f odo ~/bin    
go build -ldflags="-w -X github.com/openshift/odo/pkg/version.GITCOMMIT=9b7cc15dc" cmd/odo/odo.go

$ make test-push-devfile-test
make: *** No rule to make target 'test-push-devfile-test'.  Stop
```

**Comment by mik-dass** [comment]
> How to test changes / Special notes to the reviewer:
automation:
run make test-push-devfile-test

@yangcao77 I think this should be ` make test-cmd-devfile-push`

File: pkg/devfile/adapters/kubernetes/utils/utils.go (hunk 3 above)
Comment (yangcao77) [reply pkg/devfile/adapters/kubernetes/utils/utils.go:159]: added unit test for env variables.

File: tests/examples/source/devfiles/nodejs/devfile-with-multiple-defaults.yaml (hunk 2 above)
Comment (yangcao77) [reply tests/examples/source/devfiles/nodejs/devfile-with-multiple-defaults.yaml:74]: added test case with env variables with spaces.

File: pkg/devfile/adapters/kubernetes/utils/utils.go (hunk 1 above)
Comment (yangcao77) [reply pkg/devfile/adapters/kubernetes/utils/utils.go]: updated.

**Comment by yangcao77** [comment]
@dharmit Sorry for the confusion, it should be `make test-cmd-devfile-push`

Also updated the manual test in the PR description if you want to try it manually.

File: pkg/devfile/adapters/kubernetes/utils/utils_test.go [hunk 4]
```diff
@@ -344,9 +352,76 @@ func TestUpdateContainersWithSupervisord(t *testing.T) {
...
+					},
+				},
+			},
+			componentType:           common.ContainerComponentType,
+			expectRunCommand:        "env1=\"value1\" && " + command,
+			isSupervisordEntrypoint: true,
+			wantErr:                 false,
+		},
+		{
+			name:       "Case: custom command with multiple environment variable",
```
Comment (mik-dass) [inline pkg/devfile/adapters/kubernetes/utils/utils_test.go:391]: please add cases for the debug commands as well

Comment (yangcao77) [reply pkg/devfile/adapters/kubernetes/utils/utils_test.go]: updated

**Comment by yangcao77** [comment]
@mik-dass Can you help add back the lgtm label? I had to rebase to pull in the fix for the flaky test.

**Comment by mik-dass** [comment]
> @mik-dass Can you help add back the lgtm label? I had to rebase to pull in the fix for the flaky test.

Sure.
/lgtm

**Pull request merged**